File: leetcode/insert_delete_getrandom_o1/golang/solution.go
package golang

import "math/rand"

type RandomizedSet struct {
	valuesMap   map[int]int
	valuesSlice []int

	// This should not be included in the Leetcode solution!
	randomSource *rand.Rand
}

func Constructor() RandomizedSet {
	return RandomizedSet{
		valuesMap:    make(map[int]int),
		valuesSlice:  make([]int, 0),
		randomSource: rand.New(rand.NewSource(1)),
	}
}

func (this *RandomizedSet) Insert(val int) bool {
	_, ok := this.valuesMap[val]
	if !ok {
		this.valuesMap[val] = len(this.valuesSlice)
		this.valuesSlice = append(this.valuesSlice, val)
	}
	return !ok
}

func (this *RandomizedSet) Remove(val int) bool {
	idx, ok := this.valuesMap[val]
	if ok {
		lastValue := this.valuesSlice[len(this.valuesSlice)-1]
		this.valuesMap[lastValue] = idx
		this.valuesSlice[idx] = lastValue
		this.valuesSlice = this.valuesSlice[:len(this.valuesSlice)-1]
		delete(this.valuesMap, val)
	}
	return ok
}

func (this *RandomizedSet) GetRandom() int {
	return this.valuesSlice[this.randomSource.Intn(len(this.valuesSlice))]
}

File: leetcode/insert_delete_getrandom_o1/python/solution.py
import random


class RandomizedSet:

    _dict: dict[int, int]
    _list: list[int]

    def __init__(self) -> None:
        self._dict = dict()
        self._list = list()

    def insert(self, val: int) -> bool:
        not_exists = val not in self._dict
        if not_exists:
            self._dict[val] = len(self._list)
            self._list.append(val)
        return not_exists

    def remove(self, val: int) -> bool:
        exists = val in self._dict
        if exists:
            toRemoveIdx = self._dict.pop(val)
            replaceValue = self._list.pop()
            if val != replaceValue:
                self._list[toRemoveIdx] = replaceValue
                self._dict[replaceValue] = toRemoveIdx
        return exists

    def getRandom(self) -> int:
        return random.choice(self._list)
